bread/contrib/triggers/apps.py

```diff
@@ -0,0 +1,60 @@
+import datetime
+
+from django.apps import AppConfig
+from django.conf import settings
+from django.utils import timezone
+
+from bread.utils import get_concrete_instance
+
+TRIGGER_PERIOD = getattr(settings, "TRIGGER_PERIOD", datetime.timedelta(hours=1))
+
+
+class TriggersConfig(AppConfig):
+    name = "bread.contrib.triggers"
+    default_auto_field = "django.db.models.BigAutoField"
+
+    def ready(self):
+
+        from celery import shared_task
+        from django.db.models.signals import post_delete, post_save
+
+        from bread.utils.celery import RepeatedTask
+
+        post_save.connect(save_handler, dispatch_uid="trigger_save")
+        post_delete.connect(delete_handler, dispatch_uid="trigger_delete")
+
+        shared_task(base=RepeatedTask, run_every=TRIGGER_PERIOD)(periodic_trigger)
+
+
+def save_handler(sender, instance, created, **kwargs):
+    datachange_trigger(sender, instance, "added" if created else "changed")
+
+
+def delete_handler(sender, instance, **kwargs):
+    datachange_trigger(sender, instance, "deleted")
+
+
+def datachange_trigger(model, instance, type):
+    from django.contrib.contenttypes.models import ContentType
+
+    from .models import DataChangeTrigger
+
+    for trigger in DataChangeTrigger.objects.filter(
+        model=ContentType.objects.get_for_model(model), type=type, enable=True
+    ):
+        if trigger.filter.queryset.filter(id=instance.id).exists():
+            get_concrete_instance(trigger.action).run(instance)
+
+
+def periodic_trigger():
+
+    from .models import DateFieldTrigger
+
+    for trigger in DateFieldTrigger.objects.filter(enable=True):
+        for instance in trigger.filter.queryset.all():
+            td = trigger.triggerdate(instance)
+            if (
+                td is not None
+                and timezone.now() <= td < timezone.now() + TRIGGER_PERIOD
```

Comment:
are we triggering the action before it's triggerdate is reached?

Reply:
what happens to the objects whose triggerdate is in the past which were not triggered for some reason?
If it is critical that each action is correctly triggered, I think this system is not very robust. We should have a way of knowing whether it was triggered on a certain instance or not, so that we can trigger the ones from the past which were missed... But maybe we should discuss this as soon as we have a concrete use case, because right now it is still quite an abstract idea

Reply:
Right, if we have a trigger which fires on 2022-01-31 but our cronjob/celery task runs every hours we got 24 triggers. So we have to process "slices of time" and trigger things per slice.

Reply:
Why are we not triggering into the past? I think it is un-intuitive for the user if a job is completed before it's time. If it's late, that's more understandeable

Reply:
I know, I have thought about this some.

The problem is that if you want to get it really right this will go down a very long rabbit hole. I think relying on the operating system (to keep the processes running) and rabbitmq/celery (for the scheduling and execution) should be enough for this kind of automated actions.

For a much more controlled way of handling certain workflows one could use ```bread.contrib.workflows```, as I have addresses a few of these kinds of problem there.

I will add the execution of actions as separate tasks though, that will at least create a trace for single failed actions.

Reply:
But need to test that first.

Reply:
Okay, can you take a look again?

Reply:
cool, thanks!